pytorch_lightning/trainer/supporters.py

```diff
@@ -205,22 +227,30 @@ def __next__(self) -> Any:
         """
         Fetches the next batch from internal dataloader and restarts
         it if necessary
-
         Returns:
             Any: the resulting batch
-
         Raises:
             StopIteration: if more then :attr:`length` batches have been returned
         """
         # Note: if self.length is `inf`, then the iterator will never stop
-        if self.counter >= self.__len__():
+        if self.counter >= self.__len__() or (self.state is not None and self.state.done):
             raise StopIteration
 
         try:
             return next(self._loader_iter)
 
         except StopIteration:
+
+            if self.state is not None:
```

Comment:
See above. Let's always create a state for consistency

Reply:
Done !